### PR TITLE
Add Validation to BankAccount + CompanyStep + better handling for Plaid

### PR DESCRIPTION
**Pull request by marcaaron**
cc @roryabraham 

### Details
Adds validation to forms in the  `BankAccount` / `CompanyStep`

### Fixed Issues
Fixes https://github.com/Expensify/Expensify/issues/167761

### Tests
1. Try to break the manual account form + company information step in various ways
2. Verify helpful error messages appear in the form of growls.

### QA Steps
No QA

### Tested On

- [x] Web
- [ ] Mobile Web
- [ ] Desktop
- [ ] iOS
- [ ] Android

### CODE REVIEW

**Comment by roryabraham** [comment]
One NABish thing I'm noticing is that hitting `Enter` doesn't submit the form.

**Comment by roryabraham** [comment]
Not sure if this is intentional, but if the password is wrong and everything else is "correct", we just display an "Unknown error" growl. Might be a better UX to just say that the password is wrong.

File: src/components/AddPlaidBankAccount.js [hunk 1]
```diff
@@ -96,12 +96,9 @@ class AddPlaidBankAccount extends React.Component {
 
     render() {
         const accounts = this.getAccounts();
-        const options = _.chain(accounts)
-            .filter(account => !account.alreadyExists)
-            .map((account, index) => ({
-                value: index, label: `${account.addressName} ${account.accountNumber}`,
-            }))
-            .value();
+        const options = _.map(accounts, (account, index) => ({
```
Comment (roryabraham) [inline src/components/AddPlaidBankAccount.js:99]: Why did you remove the filter?

File: src/pages/ReimbursementAccount/BankAccountStep.js [hunk 2]
```diff
@@ -123,14 +123,16 @@ class BankAccountStep extends React.Component {
...
                             </Text>
-                            <MenuItem
-                                icon={Bank}
-                                title={this.props.translate('bankAccount.logIntoYourBank')}
-                                onPress={() => {
-                                    this.setState({bankAccountAddMethod: CONST.BANK_ACCOUNT.SETUP_TYPE.PLAID});
-                                }}
-                                shouldShowRightIcon
-                            />
+                            {!this.props.isPlaidDisabled && (
```
Comment (roryabraham) [inline src/pages/ReimbursementAccount/BankAccountStep.js:126]: Not well-versed in this topic, but if Plaid is disabled is there anything we can do for the customer? Should we disable this menu item and have a tooltip like "Reach out to concierge to re-enable this" ?

File: src/pages/ReimbursementAccount/CompanyStep.js [hunk 3]
```diff
@@ -28,24 +30,91 @@ class CompanyStep extends React.Component {
...
         };
     }
 
+    /**
+     * Validating that this is a valid address (PO boxes are not allowed)
+     *
+     * @param {String} value
+     * @returns {Boolean}
+     */
+    isValidAddress(value) {
```
Comment (roryabraham) [inline src/pages/ReimbursementAccount/CompanyStep.js:53]: I would have put some of these functions in a new lib: `src/libs/ValidationUtils.js`

File: src/pages/ReimbursementAccount/CompanyStep.js [hunk 4]
```diff
@@ -28,24 +30,91 @@ class CompanyStep extends React.Component {
...
+    /**
+     * @param {String} code
+     * @returns {Boolean}
+     */
+    validateIndustryCode(code) {
+        return !CONST.REGEX.INDUSTRY_CODE.test(code);
+    }
+
     validate() {
         // @TODO check more than just the password
```
Comment (roryabraham) [inline src/pages/ReimbursementAccount/CompanyStep.js:80]: No need for the TODO anymore

**Review by roryabraham — DISMISSED** [review]
Great work as always! I had a few suggestions, but no blockers (I guess except removing the TODO). Up to you if you want to act on any of the other suggestions.

**Comment by nickmurray47** [comment]
reviewing in a minute! sorry, have been on a long candidate call this morning

File: src/components/AddPlaidBankAccount.js (hunk 1 above)
Comment (marcaaron) [reply src/components/AddPlaidBankAccount.js:99]: The filter was just moved higher up the chain for now. 

https://github.com/Expensify/Expensify.cash/blob/125dbf3c455b37913c702bb5eb5c792bc30080ed/src/libs/actions/BankAccounts.js#L93-L94

But we are not showing any "already set up" accounts. At least for now. So, a user would need to delete that bank account in order to see those as options. There might be a better way to deal with this, but I couldn't think of one.

File: src/pages/ReimbursementAccount/BankAccountStep.js (hunk 2 above)
Comment (marcaaron) [reply src/pages/ReimbursementAccount/BankAccountStep.js:126]: Great catch as I just forgot about this :0

We have some existing copy for this but it didn't make it into this PR

>Due to a high number of login attempts, this option has been temporarily disabled for 24 hours. Please try again later or manually enter details instead.

So I'll add that now instead of disabling the option completely

**Comment by marcaaron** [comment]
Thanks for the review @roryabraham I'm looking into these two now

- [ ] The password is wrong and everything else is "correct", we just display an "Unknown error" growl. Might be a better UX to just say that the password is wrong.
- [ ] Enter doesn't submit the form.

**Comment by marcaaron** [comment]
@roryabraham can you clarify how you got to the "Unknown error"? I'm seeing this ->



So I think maybe you are running into an unknown error 😅

**Comment by marcaaron** [comment]
Ready for another review.

**Comment by nickmurray47** [comment]
noticed a few things from the old e.com bank account setup:

Screen Shot 2021-06-21 at 2 09 47 PM

Doesn't look like we have phone number validation for the `CompanyStep` right now - is that by choice?
I also like the placeholder that we had for the date format and the dropdown for the industry classification code (although I'm guessing that will prolly be a later polish step).

**Comment by marcaaron** [comment]
>Doesn't look like we have phone number validation for the CompanyStep right now - is that by choice?

@nickmurray47 Mainly because they are also caught and thrown in PHP [e.g. the phone is checked here](https://github.com/Expensify/Web-Secure/blob/bdccc8e969e8d1c79069d8add1b658cbcf23a326/lib/BankAccountAPI.php#L995-L997)

>I also like the placeholder that we had for the date format 

The placeholder is a great idea since it's not obvious what format we are preferring for that one so I'll add it now! Thanks!

>and the dropdown for the industry classification code (although I'm guessing that will prolly be a later polish step)

The industry code selector is not in scope for this project, but there are plans to eventually migrate that over.

**Comment by nickmurray47** [comment]
sounds good! 

also, was running into the same error that @roryabraham pointed out. Not sure if you or Rory's already seen this but I _think_ it's because we throw an error from secure if a user has too many errors in 24 hours, [`throwIfThrottled`](https://github.com/Expensify/Web-Secure/blob/2e018c233a8ccc0e251114f086733988491b9499/lib/BankAccount/VerificationAPI.php#L82). I noticed my account has the `throttle` NVP set and I can't progress further. Not sure if this is worthy of addressing now.

**Comment by marcaaron** [comment]
>also, was running into the same error that @roryabraham pointed out. Not sure if you or Rory's already seen this but I think it's because we throw an error from secure if a user has too many errors in 24 hours, throwIfThrottled. I noticed my account has the throttle NVP set and I can't progress further. Not sure if this is worthy of addressing now.

Hmm I'd actually expect us to be handling that here...

https://github.com/Expensify/Expensify.cash/blob/c1536a83ef628f27dcae5d0320bfba16d7bff58e/src/pages/ReimbursementAccount/ReimbursementAccountPage.js#L122-L136

I think we should maybe not block on this, but I'll create a follow up to verify that's all working correctly.

**Comment by marcaaron** [comment]
Alright so I tried to look into this comment here -> https://github.com/Expensify/Expensify.cash/pull/3686#issuecomment-865371838

And set that NVP + saw an error message. If that's your issue @nickmurray47 then you should be seeing this now:



Can you verify with me whether you see that the next time you open the modal?

**Comment by nickmurray47** [comment]
hmmm it seems like I'm still getting this generic error message. Perhaps I'm holding something wrong.

Screen Shot 2021-06-22 at 9 23 25 AM

**Comment by marcaaron** [comment]
Ok, so there's not a ton for me to go off of here... but I can suggest we try to debug (or just `Log::info()` something for) [these lines](https://github.com/Expensify/Web-Secure/blob/bdccc8e969e8d1c79069d8add1b658cbcf23a326/lib/Error/BankAccount/AutoVerifyFailure.php#L313-L323) maybe we can figure out which error is getting hit and why?

Are y'all using the recommended [test credentials for setting up a bank account](https://stackoverflow.com/c/expensify/questions/342)?

@roryabraham any thoughts on this? This isn't happening for me so I'm inclined to move forward with this.

**Comment by nickmurray47** [comment]
> Are y'all using the recommended test credentials for setting up a bank account?

Using the test credentials for the OPEN bank account. I also agree we can move forward with the PR and circle back to this error.

**Review by nickmurray47 — APPROVED** [review]

**Comment by marcaaron** [comment]
>Using the test credentials for the OPEN bank account.

Ok thanks, I just ran through again with that flow and didn't see any errors.

**Review by roryabraham — APPROVED** [review]
Don't think we need to block on this – clearly it's an edge case and we don't need perfection right now.
